feature: Add Cashfree Payment Provider

AyushChothe opened the pull request:
## Roadmap Task

👉  https://getlago.canny.io/feature-requests/p/payment-provider-add-integration-for-cashfree-payments

## Context

Cashfree Payments is India's leading payment gateway, and with Stripe halting the onboarding of new merchants in India due to regulatory constraints, businesses are left with limited alternatives. Ensuring that Cashfree Payments is available to users is essential, as it will expand Lago's accessibility and appeal to Indian companies, broadening its reach within this crucial market.

We at Pyrite Cloud have begun developing the integration with Cashfree Payments. This will help expand our platform's accessibility, especially for Indian businesses, by providing them with a reliable payment gateway as an alternative to Stripe.

## Description

Added Cashfree Payment Provider concerning existing Payment Providers. Still Work in Progress.

vincent-pochet commented:
Thank you very much @AyushChothe for this great contribution.
We will review it carefully in the coming days and we will come back to you

AyushChothe commented:
Hi @vincent-pochet,

Thank you! Please keep me updated on the review process and let me know if any changes are needed.

Also, could you guide me on how to run `Rspec` locally? I want to get started with writing tests.

vincent-pochet commented:
@AyushChothe

To run rspec locally:
- If you have a running lago api docker container (see https://github.com/getlago/lago/wiki/Development-Environment for more details):
```bash
lago exec api rspec
# or docker compose -f ./docker-compose.dev.yml exec api rspec 
```

- In you are running it directly on your local system:
```bash
cd api
bundle exec rspec
```

AyushChothe commented on config/environments/development.rb:
```diff
@@ -53,6 +53,7 @@
   config.hosts << 'api.lago.dev'
   config.hosts << 'api'
   config.hosts << 'lago.ngrok.dev'
+  config.hosts << "funky-trout-partly.ngrok-free.app"
```
Yes, I used that for testing 😅

AyushChothe commented on app/services/payment_provider_customers/create_service.rb:
```diff
@@ -65,7 +65,7 @@ def create_customer_on_provider_service(async)
 
         PaymentProviderCustomers::AdyenCreateJob.perform_now(result.provider_customer)
-      else
+      elsif result.provider_customer.type == 'PaymentProviderCustomers::GocardlessCustomer'
```
Cashfree currently does not effectively manage customers as it only creates and updates records based on the phone number as the ID. Therefore, if we implement this, the phone field must be made required in the Lago form. I'm currently focusing on the `PaymentLink` feature of Cashfree in this PR.

Ref: https://docs.cashfree.com/reference/pgcreatecustomer

AyushChothe commented:
Hey @vincent-pochet,

I've addressed all the review comments you mentioned earlier and added test cases for the impacted modules. 

I wanted to get your thoughts on how we should handle these modules since we aren't creating `PaymentProviderCustomers`. Would creating a dummy `PaymentProviderCustomers` make sense, or is there a better approach?

Here's the list of modules in question:
```ruby
PaymentProviderCustomers::Factory
PaymentProviderCustomers::CreateService
Customers::GenerateCheckoutUrlService
```

I am looking forward to hearing your thoughts!

vincent-pochet commented:
> Hey @vincent-pochet,
> 
> I've addressed all the review comments you mentioned earlier and added test cases for the impacted modules.
> 
> I wanted to get your thoughts on how we should handle these modules since we aren't creating `PaymentProviderCustomers`. Would creating a dummy `PaymentProviderCustomers` make sense, or is there a better approach?
> 
> Here's the list of modules in question:
> 
> ```ruby
> PaymentProviderCustomers::Factory
> PaymentProviderCustomers::CreateService
> Customers::GenerateCheckoutUrlService
> ```
> 
> I am looking forward to hearing your thoughts!

Thank you for the update.
To be fully compatible with the other providers, I think yes we should handle a Customer as well event if it is not attached to any external resource

AyushChothe commented:
> > Hey @vincent-pochet,
> > I've addressed all the review comments you mentioned earlier and added test cases for the impacted modules.
> > I wanted to get your thoughts on how we should handle these modules since we aren't creating `PaymentProviderCustomers`. Would creating a dummy `PaymentProviderCustomers` make sense, or is there a better approach?
> > Here's the list of modules in question:
> > ```ruby
> > PaymentProviderCustomers::Factory
> > PaymentProviderCustomers::CreateService
> > Customers::GenerateCheckoutUrlService
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I am looking forward to hearing your thoughts!
> 
> Thank you for the update. To be fully compatible with the other providers, I think yes we should handle a Customer as well event if it is not attached to any external resource

Thanks, that's what I was thinking as well. I'll start working on it now!

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,230 @@
+
+      def create_post_request
+        url = URI(::PaymentProviders::CashfreeProvider::BASE_URL)
+        request = Net::HTTP::Post.new(url)
```
An HTTP client is available in the repository to make this a bit easier:

```ruby
client = LagoHttpClient::Client.new(::PaymentProviders::CashfreeProvider::BASE_URL)
response = client.post(payment_url_params, {
  "accept" => 'application/json',
  "x-client-id" => cashfree_payment_provider.client_id,
  "x-client-secret" => cashfree_payment_provider.client_secret,
  "x-api-version" => ::PaymentProviders::CashfreeProvider::API_VERSION
})
```

vincent-pochet reviewed (commented):
It start to look great!
Here a few other minor comments

AyushChothe commented on spec/fixtures/cashfree/event.json:
```diff
@@ -0,0 +1 @@
+{"data":{"cf_link_id":1576977,"link_id":"payment_ps11","link_status":"PAID","link_currency":"INR","link_amount":"200.12","link_amount_paid":"55.00","link_partial_payments":true,"link_minimum_partial_amount":"11.00","link_purpose":"Payment for order 10","link_created_at":"2021-08-18T07:13:41","customer_details":{"customer_phone":"9000000000","customer_email":"john@gmail.com","customer_name":"John  ...
```
Actually the event is required in this compact json format for cashfree to generate and validate webhook signature.

vincent-pochet commented on app/services/payment_provider_customers/cashfree_service.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    def generate_checkout_url(send_webhook: true)
+      return result.not_found_failure!(resource: 'cashfree_payment_provider') unless cashfree_payment_provider
+      result.service_failure!(code: 'generate_checkout_url_error', message: 'Cashfree payment provider does not support the checkout url feature')
```
Since the feature is not supported, maybe you could just return an empty result, rather than a `service_failure`, what do you think?

AyushChothe replied:
Yes, we can return an empty result, but how will the user know that the feature is not supported? The user might consider it a bug.

vincent-pochet replied:
Yes I see. It will be exposed only in the APIs

`service_failre` is not supposed to be returned to the end users as it does not have a proper formatter in both REST and GraphQL APIs.

You could use `result.not_allowed_failure(code: 'feature_not_supported')` that way, the APIs will render a proper `HTTP 405` error.

vincent-pochet commented on app/services/payment_provider_customers/cashfree_service.rb:
```diff
@@ -0,0 +1,53 @@
+
+    attr_accessor :cashfree_customer
+
+    delegate :customer, to: :cashfree_customer
+
+    def cashfree_payment_provider
+      @cashfree_payment_provider ||= payment_provider(customer)
+    end
+
+    def deliver_error_webhook(cashfree_error)
+      SendWebhookJob.perform_later(
+        'customer.payment_provider_error',
+        customer,
+        provider_error: {
+          message: cashfree_error.error_message,
+          error_code: cashfree_error.code
+        }
+      )
+    end
```
It looks like this code is not in use here, so I think you can safely remove it

AyushChothe replied:
These functions are used above in `generate_checkout_url`.

vincent-pochet replied:
Since the error is expected, I don't thing there is a need for the `rescue` and the `deliver_error_webhook`  in the `generate_checkout_url` method.
Otherwise it will just spam the organization's email, without real way to address the "issue"

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,216 @@
+
+        increment_payment_attempts
+
+        # res = create_cashfree_payment
```
It looks like this file requires a bit of cleanup as well. Could you remove the code that is not in use?

vincent-pochet reviewed (commented):
Thank you for the update. Here are few small comments but I think we are close to something that could be merged :+1:

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,202 @@
+      end
+
+      def payment_url_params
+        # TODO Add settings_accessors if required
```
Does it need to be addressed?

AyushChothe replied:
No, this is not required as all the payment handling should be managed through Lago. The Payment URLs are short-lived, so the settings I considered don't make sense.

vincent-pochet replied:
Ok great! So I guess you can get rid of the comment

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,202 @@
+
+        Rails.logger.debug "invoice.total_amount_cents 2: #{invoice.total_amount_cents}"
+
+        increment_payment_attempts
+
```
With other payment provider, this service emits/creates the payment attempt via the service provider API. It looks like this one is only creating a `Payment` resource in Lago database. Is it something that is missing or does it behave differently?

AyushChothe replied:
I referred to `GoCardless` and `Adyen` providers, they create the `Payment` in the `create` method the same way and then update the status with the `update_payment_status` when the webhook is captured. Am I misunderstanding something?

vincent-pochet replied:
For example, in the Gocardless service, we are first emitting a payment on the service (https://github.com/getlago/lago-api/blob/main/app/services/invoices/payments/gocardless_service.rb#L43C29-L43C54) and then creating the payment inside Lago's database with `provider_payment_id: gocardless_result.id`. In this case `gocardless_result.id` is the ID of the payment inside gocardless database

AyushChothe replied:
As I mentioned earlier, I'm working on implementing the `Payment Link` feature for `Cashfree Payments`. Therefore, the payment emission process for Cashfree does not apply in this scenario. I create the `Payment` when the `create` event occurs and then update the status when I receive a webhook. Given this scenario, what do you think I should do?

vincent-pochet replied:
Ok, thank you for the clarification @AyushChothe.
In that case I guess the current approach should work. Could you maybe add some comments in the code to explain the workflow?

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,202 @@
+      end
+
+      def update_payment_status(provider_payment_id:, status:, metadata: {})
+        payment = if metadata[:payment_type] == 'one-time'
```
With stripe integration, this metadata value is passed when creating the payment on stripe. Does it apply the same way here?

AyushChothe replied:
Yes, you are correct. I am not using the `metadata` because I am creating the `Payment` object using the `create` method. Therefore, I can safely remove it along with the `create_payment` method.

vincent-pochet replied:
This method is used to update the status of the payment inside Lago's database when receiving a webhook from the third party service. It means that the format of the metadata even its presence will depend of each service.

vincent-pochet commented on app/services/invoices/payments/cashfree_service.rb:
```diff
@@ -0,0 +1,202 @@
+      end
+
+      def create_post_request(body)
+        client.post_with_response(body, {
```
It looks like you have to initialize the client with the service URL somewhere

AyushChothe replied:
Could you please explain this? I didn't understand it.

vincent-pochet replied:
Of sorry, I missed the method right before this one. Ignore this comment, everything looks fine here :)

vincent-pochet reviewed (commented):
Last comments regarding the `Invoices::Payments::CashfreeService` class :)

vincent-pochet commented:
It looks good to me @AyushChothe! Thank you again for this great contribution :pray: 

Someone on the front team will review the [pull request on the lago-front repository](https://github.com/getlago/lago-front/pull/1720)

In the meantime you might have to resync/rebase your branch on the `main` one of this repository.

And finally, to fix the failing `Run rails migrations` check, you can run the following command:
```bash
docker-compose -f $LAGO_PATH/docker-compose.dev.yml exec api rails annotate_models
```

vincent-pochet commented:
Closing this PR in favor of https://github.com/getlago/lago-api/pull/2767 to perform a QA and ease fixes on Lago's side
